File: python/cdp/actions/solana/wait_for_transaction_confirmation.py
from cdp.actions.solana.types import (
    WaitForTransactionConfirmationOptions,
    WaitForTransactionConfirmationResult,
)
from cdp.actions.solana.utils import get_or_create_connection


async def wait_for_transaction_confirmation(
    options: WaitForTransactionConfirmationOptions,
) -> WaitForTransactionConfirmationResult:
    """Wait for a transaction to be confirmed.

    Args:
        options: The options for waiting for a transaction to be confirmed

    Returns:
        The result of waiting for a transaction to be confirmed

    """
    connection = get_or_create_connection(options.network)

    blockhash, last_valid_block_height = await connection.get_latest_blockhash()

    confirmation = await connection.confirm_transaction(
        {
            "signature": options.signature,
            "blockhash": blockhash,
            "last_valid_block_height": last_valid_block_height,
        }
    )

    if confirmation.value.err:
        raise Exception(f"Transaction failed: {confirmation.value.err.toString()}")

    return {"signature": options.signature}

File: typescript/src/actions/solana/waitForTransactionConfirmation.ts
import { Connection } from "@solana/web3.js";

import { getOrCreateConnection, type Network } from "./utils.js";

export interface WaitForTransactionConfirmationOptions {
  /**
   * The signature of the transaction to wait for
   */
  signature: string;
  /**
   * The network or connection to use
   */
  network: Network | Connection;
}

export interface ConfirmationResult {
  /**
   * The signature of the transaction
   */
  signature: string;
}

/**
 * Wait for a transaction to be confirmed on the Solana blockchain
 *
 * @param options - The options for the transaction confirmation
 * @param options.signature - The signature of the transaction to wait for
 * @throws {Error} If the transaction fails to confirm
 *
 * @returns The confirmation result
 */
export async function waitForTransactionConfirmation(
  options: WaitForTransactionConfirmationOptions,
): Promise<ConfirmationResult> {
  const connection = getOrCreateConnection({ networkOrConnection: options.network });

  const { blockhash, lastValidBlockHeight } = await connection.getLatestBlockhash();

  const confirmation = await connection.confirmTransaction({
    signature: options.signature,
    blockhash,
    lastValidBlockHeight,
  });

  if (confirmation.value.err) {
    throw new Error(`Transaction failed: ${confirmation.value.err.toString()}`);
  }

  return { signature: options.signature };
}
